### DISCUSSION TITLE
simple example for creating a scene file

### DISCUSSION THREAD

**Discussion by avispi** [cdcseacave/openMVS#752] [Q&A]
Hello.

I have pinhole models with photos.
I want to create a scene file so I can work with openmvs.

I've used the (very impressive) openmvg-openmvs pipeline.
I'm having trouble migrating the mvg2mvs to work with my data.

Is there a simple code example somewhere?
I have several K matrices.
Every image has R and C matrices and reference to one of the K matrices.

Thanks

**Comment by cdcseacave** [comment]
all `Interface*` binaries are an example of how to use the `Interface.h` header in your project

do you have a simple data you need to convert as an example?

**Comment by avispi** [comment]
HI. Thanks so mush for offering to help.
Took me some time to arrange the data for easy readability.
I'm attaching a file.
first part contains the images extrinsic data, intrinsic data and distortion data.
distortion can be ignored. I will undistort the images.
second part contains control points using the image ids from the first part.
the control points don't have the 3d coordinate updated as I'm not using it for the bundle adjustment.
If it is necessary, I can intersect the rays from the tie-points and have the 3d coordinate updated.

link to the file:
https://bit.ly/3tIVgUQ

again- Thanks a lot for the help.

**Reply by cdcseacave** [reply]
I do not have access to the file;
pls add also the code you put in place and does not work; it will be faster to fix it

**Comment by avispi** [comment]
Hi.
I'm attaching a csv file describing my system's state after bundle adjustment and a txt file with the c++ code to create a scene.mvs

in the csv:
first part contains the image with camera exterior + interior + distortion data
second part contains control points (this is used later as the sparse cloud)

[convert_to_mvs.txt](https://github.com/cdcseacave/openMVS/files/7944276/convert_to_mvs.txt)
[block.csv](https://github.com/cdcseacave/openMVS/files/7944277/block.csv)

Again- Thanks a lot!

**Comment by cdcseacave** [comment]
pls find the code that imports your data to the OpenMVS format

some suggestions:
 - increase the precision of the rotation matrix, currently is very inaccurate with only 2 decimals
 - the sparse points do not seem to be valid, check you compute them correctly

![image](https://user-images.githubusercontent.com/2802670/151329912-c2c346df-6d4b-4883-bf84-aa447137e8c2.png)

Here is the code you can directly compile. Just call `cmake ..` in the `build` folder.
[interface.zip](https://github.com/cdcseacave/openMVS/files/7949209/interface.zip)

**Reply by avispi** [reply]
Thanks a lot.

currently, I'm not using a sparse cloud. I use tie points for the bundle adjustment equations and then I run sgm.
I saw openmvs requires a sparse cloud so I generate it using the tie points.

I'll examine the cloud and the rotation matrices.

**Reply by avispi** [reply]
Hello.
I tried another xml.
It failes to create the deapth map.
I cant understand the reason.

I'm attaching the densify step's log.
[DensifyPointCloud-220206232143001F29.log](https://github.com/cdcseacave/openMVS/files/8011301/DensifyPointCloud-220206232143001F29.log)

The picture above, in the last comment, is the sparse cloud from my input?
Can I use the opensfm tools to visualize the mvs.scene data so I can also debug my inputs?

Thank you.
Avi.

**Reply by avispi** [reply]
the data I want OPENMVS to work on consists of camera models + list of ground points where each ground point contains a list of {image_id + pixel}
this the data in the csv I added.

I use it to fill the interface object.
first I fill al the camera models.
Then, when I'm filling the point cloud.
for each vertex v I fill the 3d coordinate v.x and for each image_id+pixel in my data I add a view with the image_id to the vertex.
I'm not using the pixels in my data when I fill the interface object.

while filling, I perform this test:
from my data - I have the pixel in each view
I get the K, R, C of the relevant camera (using the image_id) 
I verify that K*R*(vertex.x-C) yields a pixel equal to the image pixel in my data.
Is this a sufficent test that all data is filled correctly?

**Reply by cdcseacave** [reply]
yes, the above sparse point cloud is from your CVS, imported into OpenMVS using the code I sent you
it is displayed with the OpenMVS Viewer

**Reply by avispi** [reply]
Hi..
Using all the help I got here, I got some nice results.
I continued working and ran into some weird behavior.

To me- it seems like all the dense point are on the edges- which maybe means I've corrupted my scene initialization and point cloud doesn't comply with the image models.

Any other thoughts?

I tried to write a sanity test to check the scene coherence.
To my understanding (because I see K matrices are normalized during scene loading from interface), all projected pixels from cloud point to corresponding views should be 0<i.j>1.
Is this correct?
And is my code also correct?

void CheckSceneSanity(const Scene& scene)
{
	auto ptItr = scene.pointcloud.points.begin();
	auto viewsItr = scene.pointcloud.pointViews.begin();
	for (int nPt = 0; nPt < scene.pointcloud.points.size(); ++nPt)
	{
		auto& point = *ptItr;
		auto& views = *viewsItr;
		for (const auto& v : views)
		{
			const auto& img = scene.images[v];
			const auto& platform = scene.platforms[img.platformID];
			const auto& cam = platform.cameras[img.cameraID];
			const auto& pose = platform.poses[img.poseID];

			cv::Mat_<double> IMinumC(3, 4);
			memset(IMinumC.data, 0, 12 * sizeof(double));
			IMinumC(0, 0) = 1;
			IMinumC(1, 1) = 1;
			IMinumC(2, 2) = 1;
			IMinumC(0, 3) = -pose.C.x;
			IMinumC(1, 3) = -pose.C.y;
			IMinumC(2, 3) = -pose.C.z;

			cv::Mat_<double> xyz(4, 1);
			xyz(0, 0) = point.x;
			xyz(1, 0) = point.y;
			xyz(2, 0) = point.z;
			xyz(3, 0) = 1;

			cv::Mat K(cam.K);
			cv::Mat R(pose.R);
			cv::Mat ij = K*R*IMinumC * xyz;
			//cv::Mat ij = platforms[v].cameras[0].K*platforms[v].poses[0].R *IMinumC;
			cv::Point3d hom(ij);
			cv::Point2d pix(hom.x / hom.z, hom.y / hom.z);
			int a = 0;
		}
	}
}

Thanks a lot!
Avi.

**Reply by cdcseacave** [reply]
not sure what is going on, and I invested already quite some time in this; pls consider making a donation so that I can allocate more time to this
